apiserver/observer/metricobserver/metricobserver.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package metricobserver
+
+import (
+	"net/http"
+	"strconv"
+	"time"
+
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/errors"
+	"github.com/juju/utils/clock"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/juju/juju/apiserver/observer"
+	"github.com/juju/juju/rpc"
+)
+
+const (
+	facadeLabel    = "facade"
+	versionLabel   = "version"
+	methodLabel    = "method"
+	errorCodeLabel = "error_code"
+)
+
+var metricLabelNames = []string{
+	facadeLabel,
+	versionLabel,
+	methodLabel,
+	errorCodeLabel,
+}
+
+// Config contains the configuration for a Observer.
+type Config struct {
+	// Clock is the clock to use for all time-related operations.
+	Clock clock.Clock
+
+	// PrometheusRegisterer is the prometheus.Registerer in which metric
+	// collectors will be registered.
+	PrometheusRegisterer prometheus.Registerer
+}
+
+// Validate validates the observer factory configuration.
+func (cfg Config) Validate() error {
+	if cfg.Clock == nil {
+		return errors.NotValidf("nil Clock")
+	}
+	if cfg.PrometheusRegisterer == nil {
+		return errors.NotValidf("nil PrometheusRegisterer")
+	}
+	return nil
+}
+
+// NewObserverFactory returns a function that, when called, returns a new
+// Observer. NewObserverFactory registers the API request metrics, and
+// each Observer updates those metrics.
+func NewObserverFactory(config Config) (observer.ObserverFactory, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Annotate(err, "validating config")
+	}
+
+	apiRequestsTotal := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "juju",
+		Subsystem: "api",
+		Name:      "requests_total",
```

Comment:
All of the metrics I've seen in Prometheus use underscores as separators. So I'm going for consistency with that.